MM-27570,MM-27757 - Activate/Deactivate plugins concurrently

mickmister opened the pull request:
#### Summary

In a [previous PR](https://github.com/mattermost/mattermost-server/pull/13617), the process of downloading plugins from the file store were was made to be concurrent. In the [discussion](https://github.com/mattermost/mattermost-server/pull/13617#discussion_r368017043) on the PR, it was thought that parallelizing the start-up process may be overkill. Another [discussion](https://mattermost.atlassian.net/browse/MM-27570?focusedCommentId=95858) on one of the tickets linked below shows that it is making cloud deployments take significantly longer when many plugins are installed.

This PR makes it so activating each plugin is done concurrently. I've also applied it to deactivating plugins for consistency and performance benefits.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-27570
https://mattermost.atlassian.net/browse/MM-27757

ali-farooq0 commented on app/plugin.go:
```diff
@@ -122,20 +128,25 @@ func (a *App) SyncPluginsActiveState() {
 
 			// Activate plugin if enabled
 			if pluginEnabled {
```
This can result in a race condition, where while a go-routine is attempting to deactivate a plugin, the second go-routine might be checking its state from `config` and end up re-activating it. I would recommend to just create two slices, one that contains a list of plugins to enable and the other one to deactivate, then you can parallelize those operations.


Race condition that might happen
```
go-routine-1 starts (line 103)
go-routine-2 starts (line 132)

go-routine-1 checks pluginEnabled state of plugin-1, starts deactivating plugin-1
go-routine-1 successfully deactivates plugin-1
go-routine-2 checks pluginEnabled state of plugin-1
go-routine-2 ends up reactivating plugin-1 
...
```

This might be mitigated if the `config.PluginStates` is a snapshot when checking states. Even if it is I would still recommend creates two slices to make it clear remove any interdependency.

mickmister replied:
> This might be mitigated if the config.PluginStates is a snapshot when checking states.

@ali-farooq0 By this, did you mean to dereference the pointer to the config at the beginning of the function?

ali-farooq0 replied:
@mickmister Nope, i meant it's already being done at the top `config := a.Config().PluginSettings`. But wasn't sure if it actually creates a snapshot or maintains a live copy that others can change underneath.

mickmister replied:
Traversing the code, `Config()` returns a pointer and does nothing underneath creates a snapshot

mickmister replied:
However, the function `a.UpdateConfig` does create a clone before editing, so there shouldn't ever be any code that is directly editing the config without being within the guard of `UpdateConfig`

mickmister commented:
@ali-farooq0 Thanks for the review, and the changes requested. I've addressed both the race condition issue per your suggested solution, and added to tests to test the `SyncPluginsActiveState` behavior when a plugin is enabled/disabled.

The way the plugin environment is started up, the only way I was able to get `env.Available()` to return this test plugin is by doing these operations in this order:

- `th := Setup(t)`
- Write plugin gzip file to `plugins` folder
    - `th.App.getBundleStorePath("testplugin")` returns `plugins/testplugin.tar.gz`
- Call `SyncPlugins`
	- Plugin is picked up by plugin environment in `plugins` folder during call to `env.Available()`
    - `InitPlugins` does not call `SyncPlugins` if the plugin env is already initialized.

I tried writing to the `plugins` folder before initializing the app, but I think something with the test setup of the app causes the `plugins` folder to be cleared. I think the way the test is laid out satisfies the testing of plugin activation/deactivation in `SyncPluginsActiveState`.

lieut-data commented on app/plugin.go:
```diff
@@ -86,46 +86,53 @@ func (a *App) SyncPluginsActiveState() {
+			go func(plugin *model.BundleInfo) {
+				defer wg.Done()
+
+				pluginId := plugin.Manifest.Id
```
nit: unneeded extra variable at this level?

mickmister commented:
> have we done measurements to confirm that this materially impacts startup performance?

@lieut-data There have not been any measurements on the solution. The only measurement I'm aware of is the 10-15 seconds it takes to start the plugins on community, taken from your comment [here](https://mattermost.atlassian.net/browse/MM-27570?focusedCommentId=95858).

> It will be interesting to see if the concurrent activation/deactivation, itself, causes downstream problems due to pre-existing assumptions about order.

The only thing that comes to mind is if a plugin mutates some structure that another plugin tries to modify concurrently, we could have race condition issues. The order itself shouldn't matter on start-up, versus something like a `MessageWillBePosted` hook where each plugin receives a potentially modified version of the post from the previous plugin.
